### PR TITLE
Update NOOP to be more CPU friendly.

### PR DESCRIPTION
**Pull request by KpjComp**
### Proposed changes in this pull request

Makes NOOP simpler by not converting it to JSON everytime. 

#### Checklist
* [ ] New tests have been added
* [x] `npm test` passes successfully
* [ ] Documentation has been updated

@amir20 to review

### CODE REVIEW

**Comment by amir20** [comment]
This build is failing because eslint found errors in your code. As the instructions suggest, you should make sure npm test works locally before pushing a pull request. 

I am not sure how familiar you are with node, jest, and pull requests. It might take a while for you to do it. If you want to learn then I suggest understanding what npm test does. 

I can't look at it today but I can close your PR and create my own someone this week. But only if you can't figure it out. I think it would be good to learn how pull requests work with tests.

**Comment by amir20** [comment]
@KpjComp See my comment above.

**Comment by KpjComp** [comment]
> I think it would be good to learn how pull requests work with tests.

Yes, it's a learning process.  :)

I can't get npm test to work, but I think it's a windows thing.
But doing  `eslint . && flow check` is now returning no errors.

Just done a new commit, but it's not appearing here but it's appearing on my fork.  Not sure if it's because it's put on hold..

**Comment by amir20** [comment]
Great, it works! Now I can actually just read code.

File: src/phantom.js [hunk 1]
```diff
@@ -37,6 +37,7 @@ const defaultLogger = createLogger();
  */
 export default class Phantom {
     logger: Logger;
+    doing_NOOP: boolean;
```
Comment (amir20) [inline src/phantom.js]: Why add this? What does it do?

File: src/phantom.js [hunk 2]
```diff
@@ -94,6 +95,12 @@ export default class Phantom {
         this.process.stdout.pipe(new Linerstream()).on('data', data => {
             const message = data.toString('utf8');
             if (message[0] === '>') {
+                //Server end has finished NOOP,
+                //lets allow NOOP again..
+                if (message === '>NOOP') {
+                    this.doing_NOOP = false;
+                    return;
+                }
```
Comment (amir20) [inline src/phantom.js]: Move this after `const json = message.substr(1)` and do `if(message === 'NOOP')` instead.

File: src/phantom.js [hunk 3]
```diff
@@ -94,6 +95,12 @@ export default class Phantom {
         this.process.stdout.pipe(new Linerstream()).on('data', data => {
             const message = data.toString('utf8');
             if (message[0] === '>') {
+                //Server end has finished NOOP,
+                //lets allow NOOP again..
+                if (message === '>NOOP') {
```
Comment (amir20) [inline src/phantom.js]: `NOOP` should be a `const` above because you are repeating it multiple places.

File: src/phantom.js [hunk 4]
```diff
@@ -315,8 +322,10 @@ export default class Phantom {
     }
 
     _heartBeat(): void {
-        if (this.commands.size === 0) {
-            this.execute('phantom', 'noop');
+        if (this.commands.size === 0 && !this.doing_NOOP) {
```
Comment (amir20) [inline src/phantom.js]: Still not sure what is the purpose of `doing_NOOP`

File: src/shim/index.js [hunk 5]
```diff
@@ -122,6 +122,18 @@ const commands = {
 function read() {
     let line = system.stdin.readLine();
     if (line) {
+        if (line === 'NOOP') {
+            //if sender has done a noop,
+            //it means sender is not busy
+            //so lets give phantom more time
+            setTimeout(function() {
```
Comment (amir20) [inline src/shim/index.js]: Why `setTimeout`?

File: src/shim/index.js [hunk 6]
```diff
@@ -122,6 +122,18 @@ const commands = {
 function read() {
     let line = system.stdin.readLine();
     if (line) {
+        if (line === 'NOOP') {
```
Comment (amir20) [inline src/shim/index.js]: If `line === NOOP` then why do you have to respond with `NOOP`. This ain't no PING PONG! :) Can you just ignore it and let the events pass through. I am pretty sure that it was doing this before.

File: src/phantom.js [hunk 7]
```diff
@@ -315,8 +322,10 @@ export default class Phantom {
...
+            this.doing_NOOP = true;
+            this.process.stdin.write('NOOP' + os.EOL, 'utf8');
```
Comment (amir20) [inline src/phantom.js]: I think the only that is required is this line and then just ignore it in `shim/index` right?

File: src/phantom.js (hunk 1 above)
Comment (KpjComp) [inline src/phantom.js]: Because your heartbeat is on a setInterval, you could end up sending a NOOP while one is still in transit / been processed.  This could cause a backlog of NOOP request's, in effect causing a memory leak.  The PhantomJS end send's an acknowledgement when it's done with the NOOP.

File: src/phantom.js (hunk 3 above)
Comment (KpjComp) [inline src/phantom.js]: Yeah, you could change to a const.  I was trying to make as little MOD's as possible.

File: src/phantom.js (hunk 2 above)
Comment (KpjComp) [inline src/phantom.js]: Well, my theory is to short circuit as early as possible.  Doing substr will just make more work for the GC, I know it's not a huge amount.  But remember these NOOP's are going 10 times a second, and was the reason I was getting CPU spike's.

File: src/phantom.js (hunk 4 above)
Comment (KpjComp) [inline src/phantom.js]: Again, stop's backlog..   I initially did it without and noticed the backlog..

File: src/shim/index.js (hunk 5 above)
Comment (KpjComp) [inline src/shim/index.js]: Ah, this is the magic bit.  In the comments it's basically mentions it.  If a NOOP request is sent, then the sender is in IDLE mode, in this mode may as well give Phantom a little bit more time.. When the sender is not busy your currently doing setTimeout(read,0), and when the sender gets busy again that's what it will use again.  Also it's only 100ms,  but it's had a massive effect on my idle CPU usage.  Like you said readLine blocks so were been extra nice to the V8 engine during idle.

File: src/phantom.js (hunk 3 above)
Comment (amir20) [inline src/phantom.js]: OK. cool. Let me know when you do.

File: src/phantom.js (hunk 2 above)
Comment (amir20) [inline src/phantom.js]: I see your point. Ok, let's keep as is.

File: src/phantom.js (hunk 4 above)
Comment (amir20) [inline src/phantom.js]: ok, good to know you saw this locally. Can you change `doing_NOOP` to something more meaningful? Like `NoopInProgress`?

File: src/shim/index.js (hunk 6 above)
Comment (amir20) [inline src/shim/index.js]: I want to understand this more. If node send `NOOP` and then phantomjs doesn't return an ACK then you are saying that it can fill backlog. I think I get what you are saying. Would it make more sense to `clearInterval` on node side instead of checking with a boolean. Because right now, it will call `hearbeat` every `100ms`. Instead, it might be more performant to wait until the ack is returned to set `setInterval` again (or `setTimeout`)?

File: src/shim/index.js (hunk 5 above)
Comment (amir20) [inline src/shim/index.js]: Interesting. I like that you tested it so I feel confident that it will be better. Also it makes sense why you need the boolean above now because of this timeout. Otherwise it would fill the queue.

**Comment by amir20** [comment]
So close! Also I will probably release this as `3.1.x` to make sure people know what the difference is.

**Comment by KpjComp** [comment]
> Instead, it might be more performant to wait until the ack is returned to set setInterval again (or setTimeout)?

That's an idea, only issue I see though is that you will end up creating 10 timers per second, instead of just doing a simple conditional check.  So not sure how V8 handles the creation & destruction of timers, may end up making more work for the GC.

> Interesting. I like that you tested it so I feel confident that it will be better.  So close! Also I will probably release this as 3.1.x

I know your busy, but before you do, could you also check / test :),  the mod's I've been doing have been on the NPM compiled code, and then I've back ported the changes.  So far I've had Zero issues, but wound't want to be responsible for breaking something.  In the mean time I'll have a look at seeing if I can create the compiled version too from source, I assume it's the gulp build, and then I can grab the lib & test.

btw. I've maybe not said it yet, but great work in starting this project in the first place.  It has really helped my use of phantom.js, and this was before these modifications.

**Comment by amir20** [comment]
> That's an idea, only issue I see though is that you will end up creating 10 timers per second, instead of just doing a simple conditional check. So not sure how V8 handles the creation & destruction of timers, may end up making more work for the GC.

What do you mean? If a `noop` is in progress then it would just clear the inveral and wait until the reponse is received. Right? Maybe I am missing something here. 

> I know your busy, but before you do, could you also check / test :), 

I'll check before merging. The only problem is all my scripts don't actually need NOOP either. I found it a long time ago while testing some edge case. I looked that phridge which did the same thing. The good news is that the code is tested 99% with coverage. So I'd feel it would work if all tests pass. 

If you can think of adding a test case then that would be awesome. But I haven't figured out how to test this. It would be really good practice to make sure all changes are tested. 

> btw. I've maybe not said it yet, but great work in starting this project in the first place. It has really helped my use of phantom.js, and this was before these modifications.

Hey, you are welcome! I have enjoyed it as well. I took over the project a couple years ago. It used to use `dnode` but that seemd to have break all the time. I kept ignoring it until I decided to rewrite everything. Then I rewrote it again using websockets. It didn't work. Then I did it again using sysin/sysout. Then a few months ago I changed everything to jest and finally I was reading yarn's code base. I leared so much from that. I added eslint every where and added flow to do static checking. I learned a lot with that and had fun with it too. V3 is pretty awesome and I hope to release v4 which target node6 going forward.

**Comment by KpjComp** [comment]
> What do you mean? If a noop is in progress then it would just clear the inveral and wait until the reponse is received. Right? Maybe I am missing something here

In more detail this is how the NOOP works, and the bit about setInterval, and more detail on the backlog check.  In my head I have how everything is working, but it can be a little tricky to explain.

Every 100ms, the setInterval checks to see if there is a message waiting to be sent too phantom.  If there is, we don't need to do anything as this will trigger the event loop in phantom anyway.  If there is no message waiting this mean's were not busy.  So if there currently is no message waiting and we have not sent a NOOP, send a NOOP now.  In aprox 100ms we should get a response from Phantom that the NOOP has been processed.  Here we could indeed clear the interval, and wait for the acknowledgement instead but it's only going to be about 100ms, and then we would need to create the setInterval again.  This process could have been simplified if Phantom had created an async readLine :(, we wound't then need a message loop / noop at all.

Now why the backlog check?.   Well on the Node side we have this NOOP check firing every 100ms,  and then on the Phantom end it detects this NOOP as an idle time, so we delay the blocking readLine by 100ms  (this makes it nicer to the V8).   Now in theory we could say there is this 100ms wait we need to do before firing another NOOP, but due to transmission time / or maybe V8 garbage collection etc, it ends up been say 120ms.  So instantly were actually already behind by 1 NOOP, node.js has sent 2 NOOP's, and Phantom has not even processed 1 yet, this will then cascade more & more NOOP's.  Now another option is maybe we could alter the Phantom end to wait say 50ms instead, and this could in theory prevent the backlog, but in theory we could still get over 100ms delay, due to the same reasons as before.  So for safety it just seems best to send an ACK.

I'll have a go at doing those Mod's tonight was out doing Squash last night, I've also got football coaching tonight, but that finishes earlier..  You any preference for what you want to re-name the boolean too, and what names do give the consts.?

**Comment by amir20** [comment]
> Here we could indeed clear the interval, and wait for the acknowledgement instead but it's only going to be about 100ms, and then we would need to create the setInterval again. This process could have been simplified if Phantom had created an async readLine :(, we wound't then need a message loop / noop at all.

I see. So you are saying it's just not worth it. Makes sense. 

> I'll have a go at doing those Mod's tonight was out doing Squash last night, I've also got football coaching tonight, but that finishes earlier.. You any preference for what you want to re-name the boolean too, and what names do give the consts.?

Let's do `const NOOP = 'noop'` and `isNoOpInProgress`.

**Comment by amir20** [comment]
@KpjComp  any update? I'd want to get to merging today :)

**Comment by amir20** [comment]
Merged in master

**Comment by KpjComp** [comment]
Sorry, @amir20  more busy this weekend than expected.   I see you've done the changes, so that's great :)
